PerspectiveBoxes/PerspectiveBoxes.js

```diff
@@ -0,0 +1,222 @@
+  const xz = line_x_xz.meet(line_z_xz);
+
+  const line_y_yz = y.join(vp_up);
+  const line_z_yz = vp_left.join(z);
+  const yz = line_y_yz.meet(line_z_yz);
+
+  // and the last lines
+  const line_x_xy = x.join(xy);
+  const line_yz_xyz = yz.join(xyz);
+  const line_xy_xyz = xy.join(xyz);
+  const line_xz_xyz = xz.join(xyz);
+
+  return {
+    corners: {
+      o,
+      x,
+      y,
+      z,
+      xy,
+      xz,
+      yz,
+      xyz,
+    },
+    edges: {
+      line_o_x,
+      line_o_y,
+      line_o_z,
+      line_x_xy,
+      line_x_xz,
+      line_y_xy,
+      line_y_yz,
+      line_z_xz,
+      line_z_yz,
+      line_xy_xyz,
+      line_xz_xyz,
+      line_yz_xyz,
+    },
+  };
+}
+
+function render_axes(cube) {
+  const { o, x, y, z } = cube.corners;
+  const x_axis = new GroupPrimitive(
+    [new LinePrimitive(o, x)],
+    Style.from_color(Color.RED).with_width(2)
+  );
+
+  const y_axis = new GroupPrimitive(
+    [new LinePrimitive(o, y)],
+    Style.from_color(Color.GREEN).with_width(2)
+  );
+
+  const z_axis = new GroupPrimitive(
+    [new LinePrimitive(o, z)],
+    Style.from_color(Color.BLUE).with_width(2)
+  );
+
+  return new GroupPrimitive([x_axis, y_axis, z_axis]);
+}
+
+function render_corners(cube) {
+  const values = Object.values(cube.corners);
+  const points = values.map((x) => new PointPrimitive(x));
+  return new GroupPrimitive(points, new Style().with_fill(Color.WHITE));
+}
+
+function render_edges(cube) {
+  const { o, x, y, z, xy, xz, yz, xyz } = cube.corners;
+
+  const lines = [
+    [o, x],
+    [o, y],
+    [o, z],
+    [x, xz],
+    [x, xy],
+    [y, xy],
+    [y, yz],
+    [z, xz],
+    [z, yz],
+    [xy, xyz],
+    [xz, xyz],
+    [yz, xyz],
+  ].map(([a, b]) => new LinePrimitive(a, b));
+
+  return new GroupPrimitive(lines, new Style().with_stroke(Color.WHITE));
+}
+
+const ANGLE = Math.PI / 3; // Math.PI / 6;
+const DISTANCE = 400;
+const CENTER = Point.point(WIDTH / 2, HEIGHT / 2);
+
+// bottom front corner of the cube
+const POINT_XY = Point.point(WIDTH / 5, (7 * HEIGHT) / 8);
+// Height from bottom front corner to top front corner
+const HEIGHT_XYZ = 600;
+// Follow the line to the left vanishing point by some percentage to
+// place the x-axis
+const PERCENT_X = 0.4;
+const VANISHING_POINTS = compute_vanishing_points(CENTER, ANGLE, DISTANCE);
+const CUBE = compute_cube(VANISHING_POINTS, POINT_XY, HEIGHT_XYZ, PERCENT_X);
+
+const VPS = new GroupPrimitive(
+  Object.values(VANISHING_POINTS).map((x) => new PointPrimitive(x)),
+  Style.from_color(Color.RED)
+);
+const CORNERS = render_corners(CUBE);
+const EDGES = render_edges(CUBE);
+const AXES = render_axes(CUBE);
+
+export const sketch = (p) => {
+  p.setup = () => {
+    p.createCanvas(
+      WIDTH,
+      HEIGHT,
+      undefined,
+      document.getElementById("sketch-canvas")
+    );
+    p.background(0);
+
+    draw_primitive(p, VPS);
+    draw_primitive(p, EDGES);
+    draw_primitive(p, CORNERS);
+    draw_primitive(p, AXES);
```

Comment:
next steps here:

- Divide the bottom and top of the cube up into sub-quads (do 16x16 for starters? or 32x32). Use the same winding order (CCW when viewed from top-down)
- Make a `Grid` of height values (in 0-1)
- Lerp between the bottom and top heights proportional to the height values (actually... that might not be correct in perspective? 🤔)
- Draw 3 quadrilaterals to connect the bottom and top quads
- Once you can do that, use multiple grids and animate!